moabb/datasets/braininvaders.py

```diff
@@ -1,39 +1,405 @@
 import glob
 import os
-import zipfile
+import os.path as osp
+import shutil
+import zipfile as z
+from distutils.dir_util import copy_tree
 
 import mne
+import numpy as np
 import yaml
 from mne.channels import make_standard_montage
+from scipy.io import loadmat
 
 from moabb.datasets import download as dl
 from moabb.datasets.base import BaseDataset
 
 
-BI2013a_URL = "https://zenodo.org/record/1494240/files/"
+BI2012a_URL = "https://zenodo.org/record/2649069/files/"
+BI2013a_URL = "https://zenodo.org/record/2669187/files/"
+BI2014a_URL = "https://zenodo.org/record/3266223/files/"
+BI2014b_URL = "https://zenodo.org/record/3267302/files/"
+BI2015a_URL = "https://zenodo.org/record/3266930/files/"
+BI2015b_URL = "https://zenodo.org/record/3268762/files/"
+
+
+def _bi_get_subject_data(ds, subject):  # noqa: C901
+    file_path_list = ds.data_path(subject)
+
+    sessions = {}
+
+    for file_path in file_path_list:
+        if ds.code in [
+            "Brain Invaders 2012",
+            "Brain Invaders 2014a",
+            "Brain Invaders 2014b",
+            "Brain Invaders 2015b",
+        ]:
+            session_name = "session_1"
+        elif ds.code == "Brain Invaders 2013a":
+            session_number = file_path.split(os.sep)[-2].replace("Session", "")
+            session_name = "session_" + session_number
+        elif ds.code == "Brain Invaders 2015a":
+            session_name = f'session_{file_path.split("_")[-1][1:2]}'
+        if session_name not in sessions.keys():
+            sessions[session_name] = {}
+
+        if ds.code == "Brain Invaders 2012":
+            condition = file_path.split("/")[-1].split(".")[0].split(os.sep)[-1]
+            run_name = "run_" + condition
+            # fmt: off
+            chnames = [
+                'F7', 'F3', 'Fz', 'F4', 'F8', 'T7', 'C3', 'Cz', 'C4',
+                'T8', 'P7', 'P3', 'Pz', 'P4', 'P8', 'O1', 'O2', 'STI 014'
+            ]
+            # fmt: on
+            chtypes = ["eeg"] * 17 + ["stim"]
+            X = loadmat(file_path)[condition].T
+            S = X[1:18, :]
+            stim = (X[18, :] + X[19, :])[None, :]
```

Comment:
Probably not necessary for this first merge to make it work with current binary P300, however if there is stimulus information available, we could keep it to be able to classify the letter / stimulus x,y location in the Braininvaders case.

Reply:
I agree, this is a good thing that the letter information are available in the data. We could use this for P300-speller classification.